src/coreclr/src/vm/interoplibinterface.cpp

```diff
@@ -0,0 +1,1249 @@
+        {
+            return &Result;
+        }
+        ExternalObjectContext* GetContext()
+        {
+            return static_cast<ExternalObjectContext*>(Result.Context);
+        }
+        ExternalObjectContext* DetachContext()
+        {
+            ExternalObjectContext* t = GetContext();
+            Result.Context = nullptr;
+            return t;
+        }
+    };
+
+    class ExtObjCxtCache
+    {
+        static Volatile<ExtObjCxtCache*> g_Instance;
+
+    public: // static
+        static ExtObjCxtCache* GetInstanceNoThrow() noexcept
+        {
+            CONTRACT(ExtObjCxtCache*)
+            {
+                NOTHROW;
+                GC_NOTRIGGER;
+                POSTCONDITION(RETVAL != NULL);
+            }
+            CONTRACT_END;
+
+            RETURN g_Instance;
+        }
+
+        static ExtObjCxtCache* GetInstance()
+        {
+            CONTRACT(ExtObjCxtCache*)
+            {
+                THROWS;
+                GC_NOTRIGGER;
+                POSTCONDITION(RETVAL != NULL);
+            }
+            CONTRACT_END;
+
+            if (g_Instance.Load() == NULL)
+            {
+                ExtObjCxtCache* instMaybe = new ExtObjCxtCache();
+
+                // Attempt to set the global instance.
+                if (NULL != FastInterlockCompareExchangePointer(&g_Instance, instMaybe, NULL))
+                    delete instMaybe;
+            }
+
+            RETURN g_Instance;
+        }
+
+    public: // Inner class definitions
+        class Traits : public DefaultSHashTraits<ExternalObjectContext *>
+        {
+        public:
+            using key_t = void*;
+            static const key_t GetKey(_In_ element_t e) { LIMITED_METHOD_CONTRACT; return (key_t)e->Identity; }
+            static count_t Hash(_In_ key_t key) { LIMITED_METHOD_CONTRACT; return (count_t)key; }
+            static bool Equals(_In_ key_t lhs, _In_ key_t rhs) { LIMITED_METHOD_CONTRACT; return (lhs == rhs); }
+        };
+
+        // Alias some useful types
+        using Element = SHash<Traits>::element_t;
+        using Iterator = SHash<Traits>::Iterator;
+
+        class LockHolder : public CrstHolder
+        {
+        public:
+            LockHolder(_In_ ExtObjCxtCache *cache)
+                : CrstHolder(&cache->_lock)
+            {
+                // This cache must be locked in Cooperative mode
+                // since releases of wrappers can occur during a GC.
+                CONTRACTL
+                {
+                    NOTHROW;
+                    GC_NOTRIGGER;
+                    MODE_COOPERATIVE;
+                }
+                CONTRACTL_END;
+            }
+        };
+
+    private:
+        friend struct InteropLibImports::RuntimeCallContext;
+        SHash<Traits> _hashMap;
+        Crst _lock;
+
+        ExtObjCxtCache()
+            : _lock(CrstExternalObjectContextCache, CRST_UNSAFE_COOPGC)
+        { }
+        ~ExtObjCxtCache() = default;
+
+    public:
+        bool IsLockHeld()
```

Comment:
Would be better to ifdef our the whole function be ifdefed for `_DEBUG` so that it does not get used by accident in release?

Reply:
That is a fair argument. I didn't realize until recently that Crst only exposed the check in `DEBUG` mode. WIll move to enclose the function itself.